File: scripts/cli-mode-e2e/anthropic-proxy.mjs
/**
 * Filtering forward-proxy (STORY-034.5 Stage 1a, Layer 1 — "blocks reaching elsewhere").
 *
 * A host-side HTTP CONNECT proxy whose allowlist is EXACTLY `api.anthropic.com:443`. The
 * cage reaches the internet ONLY through this proxy (via HTTPS_PROXY); every other host is
 * refused. It TUNNELS TLS (CONNECT) so it never sees the request body or the OAuth token —
 * it only sees the destination host:port, which it logs for audit. Per the human decision,
 * Layer 1 is "set and use" (not re-proven); Layer 2 (cage has no secrets) is the backstop.
 *
 * Usage: node anthropic-proxy.mjs [port]   (default 8889). Prints "PROXY_LISTENING <port>".
 */
import http from 'node:http';
import net from 'node:net';

const PORT = Number(process.argv[2] || 8889);
const ALLOW = new Set(['api.anthropic.com:443', 'api.anthropic.com:80']);

const log = (verb, hostport, decision) =>
  process.stderr.write(`[proxy] ${verb} ${hostport} -> ${decision}\n`);

const server = http.createServer((req, res) => {
  // Plain HTTP is not used by Claude (HTTPS only); refuse to be safe.
  log('HTTP', req.headers.host || '?', 'DENIED (https-only proxy)');
  res.writeHead(403, { 'content-type': 'text/plain' });
  res.end('forbidden: this proxy only tunnels api.anthropic.com over CONNECT\n');
});

// HTTPS tunneling via CONNECT. host:port is the target; the body stays TLS-encrypted.
server.on('connect', (req, clientSocket, head) => {
  const target = req.url || '';
  if (!ALLOW.has(target)) {
    log('CONNECT', target, 'DENIED (not on allowlist api.anthropic.com)');
    clientSocket.write('HTTP/1.1 403 Forbidden\r\n\r\n');
    clientSocket.end();
    return;
  }
  const [host, portStr] = target.split(':');
  const port = Number(portStr) || 443;
  log('CONNECT', target, 'ALLOWED');
  const upstream = net.connect(port, host, () => {
    clientSocket.write('HTTP/1.1 200 Connection Established\r\n\r\n');
    if (head && head.length) upstream.write(head);
    upstream.pipe(clientSocket);
    clientSocket.pipe(upstream);
  });
  upstream.on('error', (e) => { log('CONNECT', target, `upstream-error ${e.code || e.message}`); clientSocket.end(); });
  clientSocket.on('error', () => upstream.end());
});

server.listen(PORT, '0.0.0.0', () => {
  process.stdout.write(`PROXY_LISTENING ${PORT}\n`);
  log('start', `0.0.0.0:${PORT}`, `allowlist=[${[...ALLOW].join(', ')}]`);
});

File: scripts/cli-mode-e2e/anthropic-proxy.go
// Static HTTP CONNECT forward-proxy (STORY-034.5 Layer-1 hardening).
//
// Allowlist is EXACTLY api.anthropic.com. Runs as a container straddling a no-gateway
// `--internal` cage network (where the cage reaches it) and a network with external route
// (so it can reach Anthropic). The cage has NO other egress, so the proxy is the only way
// out — and it refuses every host except the allowlist. Built CGO-free → a static binary
// in a FROM-scratch image (no pull). It tunnels TLS via CONNECT, so it never sees the token.
package main

import (
	"io"
	"log"
	"net"
	"net/http"
	"os"
	"time"
)

var allow = map[string]bool{
	"api.anthropic.com:443": true,
	"api.anthropic.com:80":  true,
}

func handleConnect(w http.ResponseWriter, r *http.Request) {
	if !allow[r.Host] {
		log.Printf("CONNECT %s -> DENIED (not on allowlist api.anthropic.com)", r.Host)
		http.Error(w, "forbidden", http.StatusForbidden)
		return
	}
	dest, err := net.DialTimeout("tcp", r.Host, 10*time.Second)
	if err != nil {
		log.Printf("CONNECT %s -> upstream error: %v", r.Host, err)
		http.Error(w, "bad gateway", http.StatusBadGateway)
		return
	}
	log.Printf("CONNECT %s -> ALLOWED", r.Host)
	hj, ok := w.(http.Hijacker)
	if !ok {
		dest.Close()
		http.Error(w, "no hijack", http.StatusInternalServerError)
		return
	}
	client, _, err := hj.Hijack()
	if err != nil {
		dest.Close()
		return
	}
	client.Write([]byte("HTTP/1.1 200 Connection Established\r\n\r\n"))
	go func() { io.Copy(dest, client); dest.Close() }()
	io.Copy(client, dest)
	client.Close()
}

func main() {
	port := "8889"
	if len(os.Args) > 1 {
		port = os.Args[1]
	}
	h := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method == http.MethodConnect {
			handleConnect(w, r)
			return
		}
		// Plain HTTP is not used by Claude (HTTPS only); refuse.
		log.Printf("HTTP %s %s -> DENIED (https-only proxy)", r.Method, r.Host)
		http.Error(w, "forbidden: CONNECT api.anthropic.com only", http.StatusForbidden)
	})
	log.Printf("PROXY_LISTENING %s allowlist=[api.anthropic.com:443]", port)
	if err := http.ListenAndServe("0.0.0.0:"+port, h); err != nil {
		log.Fatal(err)
	}
}
